packages/webpack-cli/src/plugins/dotenv-webpack-plugin.ts

```
@@ -0,0 +1,210 @@
+        const data = this.#formatData({
+          variables,
+          target: target,
+        });
+        new DefinePlugin(data).apply(compiler);
+        await cache.storePromise("dotenv-webpack-plugin-data", null, data);
+        this.#cache = await cache.getPromise("dotenv-webpack-plugin-data", null);
+      }
+    });
+  }
+
+  #gatherVariables(compilation: any): EnvVariables {
+    const { allowEmptyValues, safe } = this.#options;
+    const vars: EnvVariables = this.#initializeVars();
+
+    const { env, blueprint } = this.#getEnvs(compilation);
+
+    Object.keys(blueprint).forEach((key) => {
+      const value = Object.prototype.hasOwnProperty.call(vars, key) ? vars[key] : env[key];
+
+      if (
+        (typeof value === "undefined" || value === null || (!allowEmptyValues && value === "")) &&
+        safe
+      ) {
+        compilation.errors.push(new Error(`Missing environment variable: ${key}`));
+      } else {
+        vars[key] = value;
+      }
+      if (safe) {
+        Object.keys(env).forEach((key) => {
+          if (!Object.prototype.hasOwnProperty.call(vars, key)) {
+            vars[key] = env[key];
+          }
+        });
+      }
+    });
+
+    return vars;
+  }
+
+  #initializeVars(): EnvVariables {
+    return this.#options.systemvars
+      ? Object.fromEntries(Object.entries(process.env).map(([key, value]) => [key, value ?? ""]))
+      : {};
+  }
+
+  #getEnvs(compilation: any): { env: EnvVariables; blueprint: EnvVariables } {
+    const { paths, safe } = this.#options;
+
+    const env: EnvVariables = {};
+    let blueprint: EnvVariables = {};
+
+    for (const path of paths) {
+      const fileContent = this.#loadFile(path, compilation);
+      Object.assign(env, dotenv.parse(fileContent));
+    }
+    blueprint = env;
+    if (safe) {
+      for (const path of paths) {
+        path.replace("[mode]", process.env.NODE_ENV || this.#compiler.options.mode || "");
```

if `process.env.NODE_ENV || this.#compiler.options.mode` is falsy, we should throw an error that we can't replace `[mode]`, and don't fallback to `""`